### DISCUSSION TITLE
(Uno.Extensions) Contribution conversation about Storage Pages

### DISCUSSION THREAD

**Discussion by DevTKSS** [unoplatform/uno#19088] [Q&A] [answered]
Added this discussion here because all requests and issues filed to that Uno.Extension topic keept unanswered there from what I seen

I did research the uno.extensions.Storage and Storage.UI Namespace in the repo and tryed to add ".UseStorage()" on different places on my app.xaml.cs file, because the docs seem to not tell how to do that. Okay, so I would have wanted to understand it from your extensions repository, looked up the ServiceCollectionExtensions.cs and HostBuilderExtensions.cs file in there and **would like to help out with Contribute to the Doc files**, to add this missing content there for all others.
**My Problem with that:**
from what I read in that mentioned files, it would mean similar approach like add it like this to the `host.Configure( ... .UseStorage((context,services) => services` but the Options visible on the lower part of this HostBuilderExtensions file:
![image](https://github.com/user-attachments/assets/71c8f02b-f48a-47c9-a3fe-df2324d7b3ef)
`.AddFileStorage` and `.AddKeyedStorage` are not available in my application to be used.

[x] UnoFeatures includes `Storage`
[x] Added the Namespace `Uno.Extensions.Storage` to my global usings (might be additionally and not needed, because its greyed out as not used
[x] tryed different places to add the .UseStorage(...) none let me see some of this Methods

Possible issue:
Although I did add the feature this only added Uno.Extensions.Storage.WinUI, but looking at the repo, there is none Namespace that is named Storage.WinUI, its Storage.UI which should be added from the uno.sdk in my opinion. but the "UseStorage" option is available in general.

Could someone check that and give me feedback if there is any issue I am making and where that can be configured? looked into the UWP Samples "FileAccess" that is not used in the uno.extensions.storage and also not in the .UI Namespace and also not added into the Packages of my app via the sdk.

**Comment by kazo0** [comment, marked as answer]
Hi @DevTKSS!

The `.UseStorage()` extension method should be automatically included as part of the `CreateDefaultBuilder()` in your App.xaml.cs. You can see this [here](https://github.com/unoplatform/uno.extensions/blob/44e66c13fcbafc6817863ccb96a875eaef8bd526/src/Uno.Extensions.Hosting.UI/UnoHost.cs#L71).

But you can always add the method yourself if you wish by setting it inside of the `.Configure(host => host`

I have attached a sample app that uses the `Storage` feature and has the `IStorage` implementation injected into the `MainModel`. We then read file info as part of a button click as can be seen by the `LoadStorageData` method in the `MainModel`.

Before clicking

<img width="409" alt="image" src="https://github.com/user-attachments/assets/8970a0f3-80b0-4902-b2ac-1608147ee966" />

After clicking

<img width="433" alt="image" src="https://github.com/user-attachments/assets/7a53277b-0f1b-4870-8c90-0aaf8d1da4c4" />


Sample app:  [UnoApp124.zip](https://github.com/user-attachments/files/18142127/UnoApp124.zip)


As for the `.AddFileStorage()` and `.AddKeyedStorage()` methods, these are internal extension methods that are not accessible from your application code. If you wish to register a different implementation of `IStorage`, you would need to provide a custom `Action<HostBuilderContext, IServiceCollection>` as part of the `configure` parameter of the `UseStorage` method

**Reply by kazo0** [reply]
Hopefully that explains a bit more! I do see that the Storage documentation does not include details concerning `UseStorage()` and using the `IStorage` interface: https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Storage/StorageOverview.html)/StorageOverview.html

I see this issue was created in the Extensions repo and you have commented on it: https://github.com/unoplatform/uno.extensions/issues/2630

If you are up for it, I would be happy to look over and review a PR from you that would add these details to the official documentation!

**Reply by DevTKSS** [reply]
Hey @kazo0 thank you for your reply!
From the "AddKeyedStorage" Method name and the passwordvault class (internal) i found in the repo files, I would have an additional question to this.
As you might know, the dotenv vault package is kind of popular in vs code using applications for example in case we use http requests. Correct me if i am wrong, but I did not see any option to store such secrets like the api key we learn about in the http and authorization extensions/features of uno. Maybe we could following up to that general storage usage for e.g. after-runtime usecase get anything working or if not do-able with that, maybe add such implementation for Enhancement? Is that realistic?
Yes, i know that there is aws and azure, but as they also told in their homepage and package description, its easyer for some bad guys to interrupt web communication than encrypt files. Also additional, it might be especially for smaller applications that are not wanting to use this services at all but want to consume data, a great thing to can use.

**Reply by DevTKSS** [reply]
@kazo0 
is there also a way to actually get the data defined in the appxmanifest file readable in the application? most likly that will be in the app.xaml.cs somewhere. for example if there would be some kind of about page in the application which is quite normal, if we could get this information automatically up to date. I think in vb.net (framework) with winforms I got there with My.Settings.Application or similar, but heared in c# we are wanting to not use reflection isnt it? would you type that information into the appsettings?

**Reply by DevTKSS** [reply]
@kazo0 can you verify, my research to be correct?
Uno is only adding the method To get the appdata nested files for storage operations and reuse the existing uwp Microsoft.extensions.storage?
In the I storage extensions interface file are more methods listed but only this one method i found with definition in the using class

**Reply by kazo0** [reply]
@DevTKSS Sorry I am having some trouble understanding what you are trying to achieve. Are you trying to use the Storage Extensions from Uno.Extensions to load and read the content of the appxmanifest file?

**Reply by DevTKSS** [reply]
@kazo0 when loading data with the extension, does that require the record to be built like IOptions (e.g. AppConfig in a new uno project with configuration enabled) or is it alliwing a full immutable structure?

**Reply by DevTKSS** [reply]
@kazo0 FYI: (re)created the docs while [this PR](https://github.com/unoplatform/uno.extensions/pull/2734). Your sample here and my current app which should load txt content showed me the current limits of Storage Extensions, what made me add this overload to this. Json and xml format is nice and easy serializeable but native formats which are not so easy to serialize slipped through this so far :) hoping it will albe accepted. But if not, the pr content might still can taken into, except the overload.

**Wanted to thank you for your support on this enhancement and closing this Discussion.**

A Storage use sample will be included in my sample app, which is building with mvux so soon as I get the state and feed things right that can be added. Thats not so easy because the docs samples and references are not relyable working "as is"😬 hopefully the src will have working sample code.

https://github.com/DevTKSS/UnoHotDesignApp1

**Comment by DevTKSS** [comment]
@kazo0 and one storage topic more:
Some KeyValueStorage definitions are public accessor and I am currently trying to get a unoapp working with dotenvx, which would be a great addition (multi environment and multi platform and encryption same as Bitcoin) to a uno app. Execution at the moment setup in csproj and first of all trying "just" with the older dotenv-vault just to get the general process of it. The problem or question I am facing is:
Does the uno.extensuions for this secrets management also import from `Environment.GetEnvironmentVariables` so I could ask this extension by providing the keyname for its vale, or would this task require me to create a extension doing this?
